File: research/src/research/observability/jsonl_log.py
"""Lightweight rolling JSONL writer for `trade_logs/<v>/...` previews.

Mirrors the Rust helper at
`server/crates/live-trader/src/jsonl_log.rs` so Python pipeline runs
write to the same agent-readable files.

- One record per line (newline-delimited JSON).
- After append, if the file has more than `MAX_LINES` lines, rewrite
  with the most recent `MAX_LINES` kept (atomic via tmp + rename).
- Heavy data still lives in DuckDB / `data/trades/` — these previews
  are for committed-to-git agent review only.
"""
from __future__ import annotations

import json
import os
import tomllib
from pathlib import Path
from typing import Any

MAX_LINES: int = 2000
DEFAULT_ROOT_ENV = "TRADE_LOGS_DIR"
DEFAULT_ROOT = Path("./trade_logs")


def repo_version() -> str:
    """Read `version` from `server/Cargo.toml` so the Python side stays
    in lockstep with the Rust workspace version. Falls back to
    `PIPFISSION_VERSION` env, then `v0.0.0` if neither is available.
    """
    env_v = os.environ.get("PIPFISSION_VERSION")
    if env_v:
        return env_v if env_v.startswith("v") else f"v{env_v}"
    here = Path(__file__).resolve()
    # research/src/research/observability/jsonl_log.py → repo at parents[4]
    candidates = [
        here.parents[4] / "server" / "Cargo.toml",
        Path.cwd() / "server" / "Cargo.toml",
    ]
    for p in candidates:
        if p.exists():
            try:
                data = tomllib.loads(p.read_text())
                v = data.get("workspace", {}).get("package", {}).get("version")
                if v:
                    return f"v{v}"
            except Exception:
                pass
    return "v0.0.0"


def trade_logs_root() -> Path:
    return Path(os.environ.get(DEFAULT_ROOT_ENV, str(DEFAULT_ROOT)))


def append(sub_path: str, record: dict[str, Any]) -> Path:
    """Append `record` as one JSONL line under
    `trade_logs/<version>/<sub_path>`. Creates parent dirs as needed.
    Returns the resolved path.
    """
    if ".." in sub_path.split("/") or sub_path.startswith("/"):
        raise ValueError(f"invalid sub_path: {sub_path!r}")
    version = repo_version()
    target = trade_logs_root() / version / sub_path
    target.parent.mkdir(parents=True, exist_ok=True)
    line = json.dumps(record, separators=(",", ":"), default=_jsonable) + "\n"
    with target.open("a", encoding="utf-8") as f:
        f.write(line)
    _roll_off_if_needed(target)
    return target


def _jsonable(o: Any) -> Any:
    """Best-effort conversion for objects json doesn't know about
    (numpy scalars, dataclasses, paths). Strings as last resort."""
    import dataclasses

    if hasattr(o, "item"):
        try:
            return o.item()
        except Exception:
            pass
    if dataclasses.is_dataclass(o):
        return dataclasses.asdict(o)
    if isinstance(o, Path):
        return str(o)
    return str(o)


def _roll_off_if_needed(path: Path) -> None:
    if not path.exists():
        return
    with path.open("r", encoding="utf-8") as f:
        lines = [ln for ln in f.readlines() if ln.strip()]
    if len(lines) <= MAX_LINES:
        return
    kept = lines[-MAX_LINES:]
    tmp = path.with_suffix(path.suffix + ".tmp")
    with tmp.open("w", encoding="utf-8") as f:
        f.writelines(kept)
    os.replace(tmp, path)

File: server/crates/live-trader/src/jsonl_log.rs
//! Lightweight rolling JSONL logger for the agent-readable
//! `trade_logs/<version>/...` previews. Each record is one line of JSON;
//! files are capped at `MAX_LINES` most-recent records and pruned in-place
//! when over the cap.
//!
//! These files are checked into git (per operator mandate) so an agent
//! can study trading behaviour without spinning up the full stack.
//! Heavy-weight forensics (full per-trade snapshots, ticks, raw bars)
//! stay outside git in `data/trades/` and `data/oanda.duckdb`.
//!
//! Atomic write semantics: append + roll uses tmp file + rename so
//! readers never see a half-written file.

use std::fs::{self, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use serde::Serialize;

/// Hard cap on lines per JSONL preview file. Older lines are dropped on
/// roll-off — the full record still exists in DuckDB / `data/trades/`.
pub const MAX_LINES: usize = 2000;

/// Default repo-relative root for committed previews. Override per call
/// when needed; everything in this module accepts an explicit root for
/// testability.
pub const DEFAULT_ROOT: &str = "trade_logs";

/// The release version that segments the on-disk layout. Pulled from the
/// workspace `version` (`env!("CARGO_PKG_VERSION")` at the call site of
/// the binary that links this crate). Kept as a free function so the
/// caller can pass either the build-time version or an env override.
pub fn version_folder(version: &str) -> String {
    if version.starts_with('v') {
        version.to_string()
    } else {
        format!("v{version}")
    }
}

/// Append `record` (any `Serialize`) as one JSONL line under
/// `<root>/<version>/<sub_path>`. Creates parent dirs as needed. After
/// append, if the file exceeds `MAX_LINES`, rewrite atomically with the
/// most recent `MAX_LINES` lines kept.
///
/// `sub_path` is a slash-joined relative path like `"EUR_USD/trades.jsonl"`
/// or `"training_log.jsonl"`. Forward slashes in instrument names
/// (`"BTC/USD"`) are flattened to `_` so OS path semantics don't create
/// phantom dirs.
pub fn append<T: Serialize>(
    root: &Path,
    version: &str,
    sub_path: &str,
    record: &T,
) -> Result<()> {
    let safe_sub = sub_path.replace("//", "/");
    // Defence-in-depth: forbid `..` segments.
    if safe_sub.split('/').any(|seg| seg == ".." || seg.is_empty()) {
        anyhow::bail!("invalid sub_path {sub_path:?}");
    }
    let folder = version_folder(version);
    let target: PathBuf = root.join(&folder).join(safe_sub);
    if let Some(parent) = target.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("creating jsonl parent {}", parent.display()))?;
    }
    let line = serde_json::to_string(record).context("serialize jsonl record")?;
    {
        let mut f = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&target)
            .with_context(|| format!("opening jsonl file {}", target.display()))?;
        f.write_all(line.as_bytes())?;
        f.write_all(b"\n")?;
    }
    roll_off_if_needed(&target)
        .with_context(|| format!("rolling off jsonl file {}", target.display()))?;
    Ok(())
}

/// If `path` has more than `MAX_LINES` lines, rewrite it with the most
/// recent `MAX_LINES` lines via tmp file + rename. No-op otherwise.
pub fn roll_off_if_needed(path: &Path) -> Result<()> {
    let f = OpenOptions::new().read(true).open(path)?;
    let reader = BufReader::new(f);
    let mut lines: Vec<String> = reader
        .lines()
        .filter_map(|r| r.ok())
        .filter(|l| !l.is_empty())
        .collect();
    if lines.len() <= MAX_LINES {
        return Ok(());
    }
    let drop_n = lines.len() - MAX_LINES;
    let kept: Vec<String> = lines.split_off(drop_n);
    let tmp = path.with_extension("jsonl.tmp");
    {
        let mut tf = OpenOptions::new()
            .create(true)
            .write(true)
            .truncate(true)
            .open(&tmp)?;
        for l in &kept {
            tf.write_all(l.as_bytes())?;
            tf.write_all(b"\n")?;
        }
        tf.flush()?;
    }
    fs::rename(&tmp, path)?;
    Ok(())
}

/// Build the resolved JSONL path for a (root, version, sub_path) triple
/// without writing. Useful for diagnostics + tests.
pub fn resolve(root: &Path, version: &str, sub_path: &str) -> PathBuf {
    root.join(version_folder(version)).join(sub_path)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use tempfile::tempdir;

    #[test]
    fn appends_and_creates_dirs() {
        let dir = tempdir().unwrap();
        append(
            dir.path(),
            "1.0.1",
            "EUR_USD/trades.jsonl",
            &json!({"a": 1, "b": "hello"}),
        )
        .unwrap();
        let p = dir.path().join("v1.0.1/EUR_USD/trades.jsonl");
        assert!(p.exists());
        let content = std::fs::read_to_string(&p).unwrap();
        assert!(content.contains("\"a\":1"));
    }

    #[test]
    fn rolls_off_at_max_lines() {
        let dir = tempdir().unwrap();
        for i in 0..(MAX_LINES + 100) {
            append(
                dir.path(),
                "v1.0.1",
                "x.jsonl",
                &json!({"i": i}),
            )
            .unwrap();
        }
        let p = dir.path().join("v1.0.1/x.jsonl");
        let content = std::fs::read_to_string(&p).unwrap();
        let n_lines = content.lines().count();
        assert_eq!(n_lines, MAX_LINES, "expected exactly MAX_LINES after roll");
        // First kept line should be index 100 (we dropped 100 oldest).
        let first: serde_json::Value =
            serde_json::from_str(content.lines().next().unwrap()).unwrap();
        assert_eq!(first["i"], 100);
    }

    #[test]
    fn rejects_traversal_sub_paths() {
        let dir = tempdir().unwrap();
        let r = append(
            dir.path(),
            "v1.0.1",
            "../escape.jsonl",
            &json!({"x":1}),
        );
        assert!(r.is_err());
    }

    #[test]
    fn version_folder_normalizes() {
        assert_eq!(version_folder("1.0.1"), "v1.0.1");
        assert_eq!(version_folder("v1.0.1"), "v1.0.1");
    }
}
